### PR TITLE
Spectators added

### PR DESCRIPTION
**Pull request by bibo38**
I've added the spectator gamemode completely to Cuberite.
This includes #1772, #2834, #2531.

Btw.: This also fixes some mobs targeting creative players(like endermen)

### CODE REVIEW

File: src/ClientHandle.cpp [hunk 1]
```diff
@@ -1532,6 +1539,27 @@ void cClientHandle::HandleSlotSelected(Int16 a_SlotNum)
 
 
 
+void cClientHandle::HandleSpectate(const AString & a_PlayerUUID)
```
Comment (worktycho) [inline src/ClientHandle.cpp]: Do we have to add another method to cClientHandle to handle this? I'm not sure this is the best long term decision.

File: src/ClientHandle.cpp [hunk 2]
```diff
@@ -1591,7 +1619,26 @@ void cClientHandle::HandleUpdateSign(
...
+			bool Item(cEntity * a_Entity) override
+			{
+				m_Player.AttachTo(a_Entity);
+				return true;
+			}
+		public:
+			cSpectateEntity(cPlayer & a_Player) : m_Player(a_Player) {}
+		} Callback(*m_Player);
+
+		m_Player->GetWorld()->DoWithEntityByID(a_TargetEntityID, Callback);
```
Comment (worktycho) [inline src/ClientHandle.cpp]: Does DoWithEntityByID not support lambdas?

File: src/ClientHandle.cpp [hunk 3]
```diff
@@ -2082,6 +2141,15 @@ void cClientHandle::SendBlockChanges(int a_ChunkX, int a_ChunkZ, const sSetBlock
 
 
 
+void cClientHandle::SendCameraSetTo(const cEntity & a_Entity)
```
Comment (worktycho) [inline src/ClientHandle.cpp]: Again, is there a good way to avoid new cClientHandle methods?

File: src/ClientHandle.cpp (hunk 1 above)
Comment (bibo38) [inline src/ClientHandle.cpp]: What's the better approach? Handeling the things in every Packet*.cpp?

File: src/ClientHandle.cpp (hunk 2 above)
Comment (bibo38) [inline src/ClientHandle.cpp]: I've always seen this kind of Callback-thing. Should I only change my callbacks or any other callbacks?

File: src/ClientHandle.cpp (hunk 3 above)
Comment (bibo38) [inline src/ClientHandle.cpp]: This are new packets... Unless you mix them with something else, but I have no idea about that...

File: src/ClientHandle.cpp (hunk 1 above)
Comment (worktycho) [inline src/ClientHandle.cpp]: No. But we need to think of a better way. The one I tried/am trying to implement is have a generic event object, which is dispatched to the relevant entities. Possibly with some form of observer/subscriber system.

File: src/ClientHandle.cpp (hunk 2 above)
Comment (worktycho) [inline src/ClientHandle.cpp]: Mostly its legacy code. We have a few hundred thousand lines of it, so updating everything to the latest practices is problematic. But lambdas are preferred for new code.

File: src/ClientHandle.cpp (hunk 1 above)
Comment (bibo38) [inline src/ClientHandle.cpp]: Wouldn't this just create for every Packet a new Class for the event?
Also it's harder to find bugs, when you use a Event queue to handle all the code.

File: src/ClientHandle.cpp (hunk 2 above)
Comment (bibo38) [inline src/ClientHandle.cpp]: Should I add the lambdas to the necessary functions in this PR?

File: src/ClientHandle.cpp (hunk 1 above)
Comment (worktycho) [inline src/ClientHandle.cpp]: Yes, which would split the code up. The problem with cClientHandle is that there is too much code in one module. And as for bugs, splitting the code up would make writing tests easier.

File: src/ClientHandle.cpp (hunk 2 above)
Comment (worktycho) [inline src/ClientHandle.cpp]: I think that would be a good idea.

**Comment by bibo38** [comment]
@madmaxoft @NiLSPACE @bearbin @tigerw Have you looked at this or can it be squashed/merged?

**Comment by bibo38** [comment]
Yeah, this was implemented in 598d90268f295e1666b2e626d3a2aabfd178b8b1. I'm just waiting for a feedback from the cuberite team.

File: src/Protocol/Protocol18x.cpp [hunk 4]
```diff
@@ -247,6 +257,16 @@ void cProtocol180::SendBlockChanges(int a_ChunkX, int a_ChunkZ, const sSetBlockV
 
 
 
+void cProtocol180::SendCameraSetTo(const cEntity & a_Entity)
```
Comment (NiLSPACE) [inline src/Protocol/Protocol18x.cpp]: This packet allows the spectator to look through the eyes of a mob right? I'm not sure if this is the right name. Perhaps something like `SendSetMobView`

Comment (bibo38) [inline src/Protocol/Protocol18x.cpp]: You're right. According to [wiki.vg](http://wiki.vg/Protocol#Camera) this packet is just named `Camera`. But be aware, that this packet works with **every entity**, e.g. you can also look through the "eyes" of a minecart. Maybe name it `SendSetCameraView` or `SendSetEntityView` if you find the current text misleading?

**Comment by LogicParrot** [comment]
Did anyone look at this?

**Comment by jammet** [comment]
As of now, spectator-clicking on a mob or player actually attacks them, no camera mode yet.

**Comment by SphinxC0re** [comment]
Any updates on this one?

**Comment by LogicParrot** [comment]
Did anyone review this?

**Comment by jammet** [comment]
Can't verify anything else at this moment, because no one else was on, today, but left clicking mobs in spectator still just attacks them instead of switching over to their camera.

**Comment by LogicParrot** [comment]
@jammet this isn't merged with master yet. Unless you've compiled yourself, you won't see the changes.

**Comment by jammet** [comment]
Ah, I see.

**Comment by bibo38** [comment]
I've updated this PR to the latest master to allow a merge of this PR.
I would be happy, if someone could test the functionality, to be sure, that I've not missed something :smirk: (just clone my [branch](https://github.com/bibo38/Cuberite/tree/spectator) and compile it).

**Comment by SphinxC0re** [comment]
Please rebase and squash your commits onto the current master

**Comment by bibo38** [comment]
@SphinxC0re Done :yum:

**Comment by mathias-github** [comment]
I will test the functionality, so we can get the PR merged as soon as possible.

**Comment by mathias-github** [comment]
These are the issues I found:
- You can't see other players who have enabled spectator mode, if you have it enabled too. Players in spectator mode should look translucent.
- When in spectator mode, your head is not translucent.
- Cannot use spectator mode's teleport feature to teleport to other players in spectator mode. (not quite sure if this is the behaviour in vanilla as well)

**Comment by mathias-github** [comment]
Before we merge this PR in the future, code for 1.9 would be good to have.

**Comment by bibo38** [comment]
yep, I've tried to get this PR merged before the 1.9 thing...
I hope I get updated this soon.

File: src/Protocol/Protocol18x.cpp [hunk 5]
```diff
@@ -2489,6 +2510,28 @@ void cProtocol180::HandlePacketSlotSelect(cByteBuffer & a_ByteBuffer)
...
+	{
+		return;
+	}
+
+	playerUUID.resize(32);
+	for (unsigned int i = 15; i < 16; i--)
+	{
+		playerUUID[i * 2 + 1] = ValueToHexDigit(playerUUID[i] & 0xf);
+		playerUUID[i * 2] = ValueToHexDigit(static_cast<UInt8>(playerUUID[i]) >> 4);
+	}
```
Comment (madmaxoft) [inline src/Protocol/Protocol18x.cpp]: This is a good candidate for a separate `cByteBuffer::ReadUUID()` function

Comment (bibo38) [inline src/Protocol/Protocol18x.cpp]: yep, you're right :yum:

**Comment by SphinxC0re** [comment]
Don't merge. Rebase.
`git fetch`
`git rebase origin/master` or `git rebase upstream/master` if that is the remote name of the canon repo. After doing that, you have to force-push: `git push --force`

**Comment by tigerw** [comment]
Why does it matter if squashing pushes it into one commit?

**Comment by LogicParrot** [comment]
It does not.

**Comment by madmaxoft** [comment]
It depends. Sometimes it makes sense to preserve the original commits, since it makes the history easier to understand in case we need to go back (bisect, blame etc.)

**Comment by LogicParrot** [comment]
>It depends. Sometimes it makes sense to preserve the original commits,

We mean that @sphinxc0re  is asking OP to squash commits, but Github can do that automatically upon merge, so that request does not matter.

**Comment by bibo38** [comment]
@LogicParrot Did I missed a Github feature? Is it not required for a PR to be manually squashed?

**Comment by sphinxc0re** [comment]
AFAIK the sqash-merge does only that: Merging. Since we want a clean history, we require everyone to rebase onto the current master

**Comment by madmaxoft** [comment]
The squash-merge does actually squash all the commits into a single one and rebases it onto the current master, so it is useful for those scenarios when we want the PR to be merged as a single commit. If we want to keep it in several commits, we still need to manually rebase and then merge.

**Comment by PureTryOut** [comment]
I assume this has been merged, or at least partially? I can succesfully run `/gamemode spectator` in-game and I can noclip fine, I'm even hidden from other players. The only issue I see is that mobs still target you, which seems to be fixed in this PR.

**Comment by jammet** [comment]
You also can't right click another player or mob to see through their eyes.

**Comment by jammet** [comment]
Disregard my last message, I thought I was responding to the top of the
thread.

**Comment by bibo38** [comment]
> ERROR: Newly undocumented API symbols found:
> cFloater.CanPickup (function)
> cFloater.GetAttachedMobID (function)
> cFloater.GetOwnerID (function)
> 
> The command "./CIbuild.sh" exited with 1.
(https://travis-ci.org/cuberite/cuberite/jobs/150280996#L2581)

Is this a CI problem? I don't see the problem with my PR.

**Comment by bearbin** [comment]
Yes, this is a CI problem.

**Comment by bibo38** [comment]
Currently everything works except the [issues reported by @mathiascode](https://github.com/cuberite/cuberite/pull/2852#issuecomment-218860366) (teleporting to other spectators is not possible in vanilla). I would be pleased if somebody knows how to fix these issues(missing packets, ...).

**Comment by mathiascode** [comment]
Since everything works fine already, maybe we should merge this as it is, and fix minor bugs later?

**Comment by madmaxoft** [comment]
Okay, I agree with a merge. @bibo38 would you mind if I squash-merged all your commits into a single one? That would be preferable to having "merged master" intermediate commits. Also, I need you to fix the merge conflict first.

**Comment by bibo38** [comment]
@madmaxoft No problem. But the merging could take some time.

**Comment by bibo38** [comment]
@madmaxoft Now it is merged to the current master. Only the ApiDesc.lua needed a manual merge.
Get it merged :wink:

**Comment by madmaxoft** [comment]
Nooo, not a merge! **Rebase**

**Pull request merged**